File: screenshot/ViewControllers/ScreenshotDisplay/ScreenshotDisplayViewController.m
//
//  ScreenshotDisplayViewController.m
//  screenshot
//
//  Created by Corey Werner on 8/16/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

#import "ScreenshotDisplayViewController.h"

#import "screenshot-Swift.h"

// Uncomment to calculate b0,b1 for a new tutorial screenshot.
#define STORE_NEW_TUTORIAL_SCREENSHOT 1

@interface ScreenshotDisplayViewController () <UIScrollViewDelegate>

@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UIImageView *screenshotImageView;
@property (nonatomic, strong) UIView *screenshotImageFrameView;
@property (nonatomic, assign) CGPoint b0;

@end

@implementation ScreenshotDisplayViewController

#pragma mark - Life Cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    CGFloat p = [Geometry padding];
    CGFloat statusBarHeight = [UIApplication sharedApplication].statusBarFrame.size.height;
    
    UIView *backgroundView = [[UIView alloc] init];
    backgroundView.translatesAutoresizingMaskIntoConstraints = NO;
    backgroundView.backgroundColor = [UIColor colorWithWhite:97.f/255.f alpha:1.f];
    [self.view addSubview:backgroundView];
    [backgroundView.topAnchor constraintEqualToAnchor:self.view.topAnchor constant:statusBarHeight].active = YES;
    [backgroundView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor].active = YES;
    [backgroundView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
    [backgroundView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor].active = YES;
    
    self.scrollView = ({
        UIScrollView *scrollView = [[UIScrollView alloc] init];
        scrollView.translatesAutoresizingMaskIntoConstraints = NO;
        scrollView.delegate = self;
        scrollView.maximumZoomScale = 3.f;
        scrollView.contentInset = UIEdgeInsetsMake(p, p, p, p);
        [self.view addSubview:scrollView];
        [scrollView.topAnchor constraintEqualToAnchor:self.topLayoutGuide.bottomAnchor].active = YES;
        [scrollView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor].active = YES;
        [scrollView.bottomAnchor constraintEqualToAnchor:self.bottomLayoutGuide.topAnchor].active = YES;
        [scrollView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor].active = YES;
        scrollView;
    });
    
    CGFloat horizontal = self.scrollView.contentInset.left + self.scrollView.contentInset.right;
    CGFloat vertical = self.scrollView.contentInset.top + self.scrollView.contentInset.bottom;
    
    UIImageView *screenshotImageView = self.screenshotImageView;
    screenshotImageView.contentMode = UIViewContentModeScaleAspectFit;
    [self.scrollView addSubview:screenshotImageView];
    screenshotImageView.translatesAutoresizingMaskIntoConstraints = NO;
    [screenshotImageView.topAnchor constraintEqualToAnchor:self.scrollView.topAnchor].active = YES;
    [screenshotImageView.leadingAnchor constraintEqualToAnchor:self.scrollView.leadingAnchor].active = YES;
    [screenshotImageView.bottomAnchor constraintEqualToAnchor:self.scrollView.bottomAnchor].active = YES;
    [screenshotImageView.trailingAnchor constraintEqualToAnchor:self.scrollView.trailingAnchor].active = YES;
    [screenshotImageView.widthAnchor constraintEqualToAnchor:self.scrollView.widthAnchor constant:-horizontal].active = YES;
    [screenshotImageView.heightAnchor constraintEqualToAnchor:self.scrollView.heightAnchor constant:-vertical].active = YES;
    
#ifdef STORE_NEW_TUTORIAL_SCREENSHOT
    screenshotImageView.userInteractionEnabled = YES;
    UILongPressGestureRecognizer *longPressGestureRecognizer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(handleLongPressGesture:)];
    [screenshotImageView addGestureRecognizer:longPressGestureRecognizer];
#endif
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    
#ifndef STORE_NEW_TUTORIAL_SCREENSHOT
    if (!CGRectIsEmpty(self.screenshotImageView.bounds)) {
        [self insertShoppableFrames];
    }
#endif
}


#pragma mark - Layout

- (CGRect)aspectFitRectForImageSized:(CGSize)imageSize inViewSized:(CGSize)viewSize {
    CGFloat imageScale = MIN(viewSize.width / imageSize.width, viewSize.height / imageSize.height);
    CGSize scaledImageSize = CGSizeMake(imageSize.width * imageScale, imageSize.height * imageScale);
    
    CGRect rect = CGRectZero;
    rect.origin.x = round((viewSize.width - scaledImageSize.width) * .5f);
    rect.origin.y = round((viewSize.height - scaledImageSize.height) * .5f);
    rect.size.width = round(scaledImageSize.width);
    rect.size.height = round(scaledImageSize.height);
    return rect;
}


#pragma mark - Image

- (UIImageView *)screenshotImageView {
    if (!_screenshotImageView) {
        _screenshotImageView = [[UIImageView alloc] init];
    }
    return _screenshotImageView;
}

- (void)setImage:(UIImage *)image {
    self.screenshotImageView.image = image;
}

- (UIImage *)image {
    return self.screenshotImageView.image;
}


#pragma mark - Shoppable

- (void)insertShoppableFrames {
    [self.screenshotImageFrameView removeFromSuperview];
    
    CGRect imageFrame = [self aspectFitRectForImageSized:self.image.size inViewSized:self.screenshotImageView.bounds.size];
    
    UIView *screenshotImageFrameView = [[UIView alloc] initWithFrame:imageFrame];
    screenshotImageFrameView.userInteractionEnabled = NO;
    [self.screenshotImageView addSubview:screenshotImageFrameView];
    self.screenshotImageFrameView = screenshotImageFrameView;
    
    for (Shoppable *shoppable in self.shoppables) {
        CGRect frame = [shoppable frameWithSize:screenshotImageFrameView.bounds.size];
        
        UIView *frameView = [[UIView alloc] initWithFrame:frame];
        frameView.layer.borderColor = [[UIColor whiteColor] colorWithAlphaComponent:.7f].CGColor;
        frameView.layer.borderWidth = 2.f;
        [screenshotImageFrameView addSubview:frameView];
    }
}


#pragma mark - Scroll View

- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
    return self.screenshotImageView;
}

#pragma mark - Gesture Recognizer

- (void)handleLongPressGesture:(UILongPressGestureRecognizer *)gestureRecognizer {
    if (gestureRecognizer.state != UIGestureRecognizerStateBegan) {
        return;
    }
    CGPoint longPressPoint = [gestureRecognizer locationInView:self.screenshotImageView];
    CGFloat normalizedX = longPressPoint.x / self.screenshotImageView.bounds.size.width;
    CGFloat normalizedY = longPressPoint.y / self.screenshotImageView.bounds.size.height;
    
//    MAX(0, normalizedX)
    if (normalizedX < 0) normalizedX = 0;
    if (normalizedY < 0) normalizedY = 0;
    if (normalizedX > 1) normalizedX = 1;
    if (normalizedY > 1) normalizedY = 1;
    CGPoint normalizedPressPoint = CGPointMake(normalizedX, normalizedY);
    if (CGPointEqualToPoint(self.b0, CGPointZero)) {
        self.b0 = normalizedPressPoint;
        NSLog(@"b0:%@  longPressPoint:%@  in size:%@", NSStringFromCGPoint(self.b0), NSStringFromCGPoint(longPressPoint), NSStringFromCGSize(self.screenshotImageView.bounds.size));
    } else {
        CGPoint b1 = normalizedPressPoint;
        NSLog(@"b1:%@  longPressPoint:%@  in size:%@", NSStringFromCGPoint(b1), NSStringFromCGPoint(longPressPoint), NSStringFromCGSize(self.screenshotImageView.bounds.size));
        CGFloat viewWidth = self.screenshotImageView.bounds.size.width;
        CGFloat viewHeight = self.screenshotImageView.bounds.size.height;
        CGRect frame = CGRectMake(self.b0.x * viewWidth, self.b0.y * viewHeight, (b1.x - self.b0.x) * viewWidth, (b1.y - self.b0.y) * viewHeight);
        UIView *frameView = [[UIView alloc] initWithFrame:frame];
        frameView.layer.borderColor = [[UIColor greenColor] colorWithAlphaComponent:.7f].CGColor;
        frameView.layer.borderWidth = 2.f;
        [self.screenshotImageView addSubview:frameView];
        self.b0 = CGPointZero;
    }
}


@end

File: screenshot/View/ScreenshotDisplay/ScreenshotDisplayViewController.swift
//
//  ScreenshotDisplayViewController.swift
//  screenshot
//
//  Created by Gershon Kagan on 2/11/18.
//  Copyright (c) 2018 crazeapp. All rights reserved.
//

import UIKit

class ScreenshotDisplayViewController: BaseViewController, UIScrollViewDelegate {
    
    // MARK: Properties
    
    let scrollView = UIScrollView()
    var screenshotImageFrameView: UIView?
    var b0: CGPoint = .zero
    
    // MARK: - Life Cycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = .gray4
        
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.delegate = self
        scrollView.maximumZoomScale = 3
        scrollView.contentInset = UIEdgeInsets(top: .padding, left: .padding, bottom: .padding, right: .padding)
        self.view.addSubview(scrollView)
        scrollView.topAnchor.constraint(equalTo:self.topLayoutGuide.bottomAnchor).isActive = true
        scrollView.leadingAnchor.constraint(equalTo:self.view.leadingAnchor).isActive = true
        scrollView.bottomAnchor.constraint(equalTo:self.bottomLayoutGuide.topAnchor).isActive = true
        scrollView.trailingAnchor.constraint(equalTo:self.view.trailingAnchor).isActive = true
        
        let horizontal: CGFloat = self.scrollView.contentInset.left + self.scrollView.contentInset.right
        let vertical: CGFloat  = self.scrollView.contentInset.top + self.scrollView.contentInset.bottom
        
        screenshotImageView.contentMode = .scaleAspectFit
        self.scrollView.addSubview(screenshotImageView)
        screenshotImageView.translatesAutoresizingMaskIntoConstraints = false
        screenshotImageView.topAnchor.constraint(equalTo:self.scrollView.topAnchor).isActive = true
        screenshotImageView.leadingAnchor.constraint(equalTo:self.scrollView.leadingAnchor).isActive = true
        screenshotImageView.bottomAnchor.constraint(equalTo:self.scrollView.bottomAnchor).isActive = true
        screenshotImageView.trailingAnchor.constraint(equalTo:self.scrollView.trailingAnchor).isActive = true
        screenshotImageView.widthAnchor.constraint(equalTo:self.scrollView.widthAnchor, constant:-horizontal).isActive = true
        screenshotImageView.heightAnchor.constraint(equalTo:self.scrollView.heightAnchor, constant:-vertical).isActive = true
        
        screenshotImageView.didLayoutSubviews = { [weak self] in
            self?.insertShoppableFrames()
        }
        
    }
    
    override var previewActionItems: [UIPreviewActionItem] {
        guard let screenshot = screenshot else {
            return []
        }
        
        let deleteAction = UIPreviewAction(title: "generic.delete".localized, style: .destructive) { (action, viewController) in
            DataModel.sharedInstance.hide(screenshotOIDArray: [screenshot.objectID], kind:.single)
        }
        return [deleteAction]
    }
    
    // MARK: Screenshot
    
    fileprivate var shoppablesFrc: FetchedResultsControllerManager<Shoppable>?
    
    var screenshot: Screenshot? {
        didSet {
            if let screenshot = screenshot {
                let shoppablesFrc = DataModel.sharedInstance.shoppableFrc(delegate: self, screenshot: screenshot)
                self.shoppablesFrc = shoppablesFrc
                
                if let data = screenshot.imageData, let image = UIImage(data: data as Data) {
                    self.image = image
                }
                
            }
        }
    }
    
    // MARK: - Image
    
    fileprivate let screenshotImageView = ImageView()
    
    fileprivate var image: UIImage? {
        get {
            return self.screenshotImageView.image
        }
        set(newImage) {
            self.screenshotImageView.image = newImage
        }
    }
    
    // MARK: - Shoppable
    
    @objc func insertShoppableFrames() {
        guard let image = image, !screenshotImageView.bounds.isEmpty else {
            print("insertShoppableFrames empty image")
            return
        }
        self.screenshotImageFrameView?.removeFromSuperview()
        
        let imageFrame = image.size.aspectFitRectInSize(self.screenshotImageView.bounds.size)
        
        let screenshotImageFrameView = UIView(frame: imageFrame)
        screenshotImageFrameView.isUserInteractionEnabled = false
        self.screenshotImageView.addSubview(screenshotImageFrameView)
        self.screenshotImageFrameView = screenshotImageFrameView

        
        if let shoppables = self.shoppablesFrc?.fetchedObjects {
            for shoppable in shoppables {
                let frame = shoppable.frame(size: screenshotImageFrameView.bounds.size)
                
                let frameView = UIView(frame: frame)
                frameView.layer.borderColor = UIColor.white.withAlphaComponent(0.7).cgColor
                frameView.layer.borderWidth = 2
                screenshotImageFrameView.addSubview(frameView)
            }
        }
    }
    
    // MARK: - Scroll View

    func viewForZooming(in scrollView: UIScrollView) -> UIView? {
        return self.screenshotImageView
    }
    
}

extension ScreenshotDisplayViewController: FetchedResultsControllerManagerDelegate {
    func managerDidChangeContent(_ controller: NSObject, change: FetchedResultsControllerManagerChange) {
        if isViewLoaded {
            insertShoppableFrames()
        }
    }
}

fileprivate extension ScreenshotDisplayViewController {
    class ImageView: UIImageView {
        var didLayoutSubviews: (()->())?
        
        override func layoutSubviews() {
            super.layoutSubviews()
            didLayoutSubviews?()
        }
    }
}
